rfcs/006-package-management.md

```diff
@@ -0,0 +1,425 @@
+---
+feature: package management
+start-date: 2024-06-28
+author: Joe Neeman
+---
+
+# Package management
+
+People want to reuse code, but Nickel doesn't currently have a good way to do
+it. We should have a way to fetch packages and make them easily available to
+Nickel code. The mechanism needs to be predictable (it should fetch the code
+that the user expects to fetch) and reliable (if it works on my machine it
+should work on your machine).
+
+Explicit non-goals are:
+
+- Management of system (or other non-Nickel) dependencies. You can use Nix for
+  that.
+- Any sort of processing or compilation. Nickel is an interpreted language, so
+  package management is only in charge of distributing unmodified source files.
+- Integration with other language ecosystems; this is only for Nickel code.
+  Anyone integrating Nickel with other languages should use a polyglot build
+  system like Bazel. (However, providing integration points for other tools *is*
+  in scope. More detail on that below.)
+
+## The manifest file
+
+We will require a manifest file in order to import packages. Manifest files
+must be named `electroplate.ncl`, and they are found by searching up from the
+file being evaluated. That is, when the user invokes `nickel export path/to/
+foo.ncl`, we look for a manifest at `path/to/electroplate.ncl` and then at
+`path/electroplate.ncl`, and so on.
+
+The manifest file format is defined by the contract `std.package.Manifest`,
+which is defined as
+
+```nickel
+{
+  name | String,
+  version | Semver,
+  nickel-version
+    | Semver,
+    | doc "The minimum version of nickel supported by this package",
+
+
+  dependencies
+    | { _:
+      [|
+        'Path String,
+        'Git { url | String, branch | optional | String, rev | optional | String },
+        'Index { name | String, version | SemverConstraint },
+      |]
+      }
+    | default
+    = {},
+}
+```
+
+So an example manifest might look like
+
+```nickel
+{
+  name = "demo",
+  version = "0.1.0",
+  nickel-version = "1.0",
+  dependencies = {
+    foo = 'Index { package = "github/tweag/foo", version = "1.2.0" },
+    bar = 'Path "../my-bar",
+  }
+} | std.package.Manifest
+```
+
+### Alternative: inline dependencies
+
+Nix and Dhall allow for importing dependencies dynamically, using things like
+`fetchGit`. There was some discussion
+[here](https://github.com/tweag/nickel/issues/329#issuecomment-967372858)
+on the advantages and disadvantages of inline imports.
+
+### Alternative: toml manifest
+
+Maybe the manifest should be in some plain-data format like toml. This would
```

Comment:
I didn't see it mentioned explicitly, so I wanted to raise it. Will `import` be forbidden in the manifest itself? I would imagine that that would be a good idea. But I haven't thought about it very deeply.

Reply:
Good question, I had originally thought that the manifest wouldn't support package management, but it would be allowed to import normal paths. But maybe let's start by forbidding imports altogether.